### PR TITLE
[Perf] Reduce Triton recompiles for multimodal attention ranges

### PR DESCRIPTION
**Pull request by dumko2001**
## Purpose

Part of #48650.

Triton unified attention receives `MAX_MM_RANGES` as a `tl.constexpr`. Passing
every exact multimodal range count creates a new specialization on first use.
Pad production metadata to the next power-of-two width with inert `[0, 0]`
rows, reducing the specialization set to 1/2/4/8/... without changing logical
ranges.

This matches vLLM's existing power-of-two specialization policy. #48734,
#48739, and #48813 own independent constexpr sites and have no file overlap.
Mandatory duplicate searches found no competing implementation.

AI assistance was used. The human submitter must review every changed line and
the recorded performance evidence before readiness.

## Implementation

- Bucket Triton production metadata before device transfer.
- Preserve the unbucketed default used by FlashAttention.
- Guard nonstandard direct unified-kernel callers, including zero-width input.
  The INT4 direct dispatcher bypasses this fallback, but its production
  metadata is already bucketed by the shared builder.
- Leave the kernel loop and invalid-range semantics unchanged.
- Test padding, dtype/device, numerical equivalence, and JIT reuse for prefill
  and decode at both 3→4 and 5→8 bucket transitions.

## Validation

Current head `d6b338142` is based on upstream `1a659a0c3`.

```text
pytest -q \
  tests/kernels/attention/test_triton_unified_attention.py::test_mm_prefix_range_count_does_not_recompile \
  --collect-only
4 tests collected

pre-commit run --files tests/kernels/attention/test_triton_unified_attention.py
all applicable hooks passed, including mypy

git diff --check
passed
```

An earlier production-equivalent revision passed the builder test and the
3→4 prefill JIT-monitor test on H100. Fresh-cache service testing on H100
compiled only widths 1/2/4/8 for logical widths 1..8, rather than eight exact
specializations. A Gemma 3 W8A8 service selected `TRITON_ATTN`, completed graph
capture, and served image counts `1,2,3,4,5,6,7,8,3,5,7` without recompiling
the repeated 3/5/7 requests.

The current test is stronger than that historical GPU run: it covers 3→4 and
5→8 on forced 2D prefill and 3D decode. It has only been collected locally;
no current-head H100 or MI300X pass is claimed.

Still required before readiness:

- Run the four-case current-head test on H100 and MI300X/gfx942.
- Measure steady-state raw-vs-padded 3/4 and 5/8 kernel cost on both platforms,
  excluding Python padding and reporting repeated distributions rather than a
  hard CI threshold.
- Compare representative model output with the parent or another backend,
  then obtain human review and full CI.

## Release note

Reduce serving-time Triton attention recompilations for multimodal batches by
grouping prefix-range specializations into power-of-two buckets.

This is a mainline performance change, not a v0.26 cherry-pick candidate.

### CODE REVIEW

File: vllm/v1/attention/ops/triton_unified_attention.py [hunk 1]
```diff
@@ -908,10 +908,17 @@ def unified_attention(
...
     if mm_prefix_range is not None:
         if mm_prefix_range.ndim == 3:
             use_mm_prefix = True
             max_mm_ranges = mm_prefix_range.shape[1]
+            max_mm_ranges_bucket = 1 << (max_mm_ranges - 1).bit_length()
+            if max_mm_ranges != max_mm_ranges_bucket:
+                mm_prefix_range = torch.nn.functional.pad(
+                    mm_prefix_range,
+                    (0, 0, 0, max_mm_ranges_bucket - max_mm_ranges),
+                )
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply vllm/v1/attention/ops/triton_unified_attention.py]: Agreed — addressed in affcbdd. The power-of-two bucketing + padding now happens once at metadata build time (before the H2D copy) in `compute_mm_prefix_range_tensor` via a new `pad_to_power_of_two=True` flag passed from the Triton backend, so the request/per-layer path no longer allocates. `unified_attention` keeps its bucketing guard as an inert fallback for direct callers, so it's now a no-op int comparison on the hot path. Scoped to the Triton backend only — flash_attn keeps the default (no bucketing). Added a CPU unit test in `tests/v1/attention/test_attention_backend_utils.py`.

**Comment by dumko2001** [comment]
Current-head CUDA validation (`f85d34f29`, H100 SXM 80 GB):

```text
pytest -q \
  tests/v1/attention/test_attention_backends.py::test_compute_mm_prefix_range_tensor_power_of_two_padding \
  tests/kernels/attention/test_triton_unified_attention.py::test_mm_prefix_range_count_does_not_recompile
2 passed, 14 warnings
```

The second test enabled vLLM's JIT monitor around the real Triton unified-attention kernel and verified that changing the logical multimodal-range count inside one power-of-two bucket does not produce another compiled specialization. The numerical assertions also confirm that padded `[0, 0]` entries preserve the three-range result.

RunPod had no MI300X inventory during this audit, so ROCm and a mixed-image service-level compile/TTFT benchmark remain unclaimed. Keeping the PR draft until those cross-platform/service checks or maintainer guidance narrow the merge gate.

**Comment by dumko2001** [comment]
Current head `9744e84d8` expands the existing JIT regression from one 3→4
prefill case to the full intended matrix:

```text
3→4 prefill/2D
3→4 decode/3D
5→8 prefill/2D
5→8 decode/3D
```

All four retain the no-recompile assertion and raw-vs-padded numerical parity.
Local collection reports four tests and all focused pre-commit hooks pass. The
production change is unchanged, but the stronger test revision has not yet run
on H100 or MI300X, so the PR remains draft pending those runs and controlled
steady-state measurements.
